### PR TITLE
feat: Added regexp_match operator support for bigquery

### PR DESCRIPTION
**Pull request by TeddyCr**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Adds support for `regexp_match` sqlalchemy operator https://docs.sqlalchemy.org/en/14/core/sqlelement.html#sqlalchemy.sql.expression.ColumnElement.regexp_match

### CODE REVIEW

File: tests/system/test_sqlalchemy_bigquery.py [hunk 1]
```diff
@@ -774,3 +774,27 @@ def test_unnest(engine, bigquery_dataset):
...
     assert sorted(r[0] for r in conn.execute(query)) == ["a", "b", "c", "x", "y"]
+
+
+@pytest.mark.skipif(
+    packaging.version.parse(sqlalchemy.__version__) < packaging.version.parse("1.4"),
+    reason="regexp_match support requires version 1.4",
+)
+def test_regexp_match(session, table):
+    results = session.query(table.c.string).where(
+        table.c.string.regexp_match(".*foo.*")
```
Comment (TeddyCr) [inline tests/system/test_sqlalchemy_bigquery.py:785]: hey folks 👋 I might need some help here for more specific regex pattern matching

**Comment by TeddyCr** [comment]
Hey folks 👋 just wanted to loop back in here. Let me know if you need anything else from my end to move forward with the PR 🙂

**Comment by TeddyCr** [comment]
Hey folks 👋 just wanted to loop back in here. Let me know if you need anything else from my end to move forward with the PR 🙂

**Comment by harshach** [comment]
@parthea can you please review this PR. We are waiting for this to be merged. If there is any feedback please let us know. Thanks

**Comment by harshach** [comment]
@alvarowolfx @chalmerlowe @parthea we really need help with merging this PR. We are essentially maintaining a fork just to get this PR into our releases. Please let us know what we need to do get this merged in. Thanks

**Comment by chalmerlowe** [comment]
@harshach I will do what I can to clear my schedule to prioritize looking at this PR this week.
Your patience is appreciated. I will provide feedback as soon as I am able to examine this item.

File: tests/system/test_sqlalchemy_bigquery.py (hunk 1 above)
Comment (chalmerlowe) [reply tests/system/test_sqlalchemy_bigquery.py:785]: @TeddyCr If I am tracking this correctly based on the following pytest fixture, it appears that the table we match up against gets generated in the `table` method and uses the data in [sample.json](https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f05979353f925b92def135f556fd267f1b71a190/tests/system/data/sample.json) 

```
@pytest.fixture(scope="session")
def table(engine, bigquery_dataset):
    return Table(f"{bigquery_dataset}.sample", MetaData(bind=engine), autoload=True)
```

It appears we are making the dataset here... 
https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f05979353f925b92def135f556fd267f1b71a190/tests/system/conftest.py#L73

using the `load_sample_data()` method:
https://github.com/googleapis/python-bigquery-sqlalchemy/blob/f05979353f925b92def135f556fd267f1b71a190/tests/system/conftest.py#L38

Long story short... in looking at the data in the string column found in the sample.json file, it appears that we have a several options for expressions to search for. For example, we could search for `Bleecker`:

`table.c.string.regexp_match(".*Bleecker.*")`

**Comment by TeddyCr** [comment]
Thanks @chalmerlowe I'll address the comment. 🙂

**Comment by TeddyCr** [comment]
@chalmerlowe just pushed the changes based on your comments. Quick question regarding `import test_utils.prefixer` and `import test_utils.retry`, are these internal packages? I was trying to install these to execute the tests on my side but could not figure out what package this is.

**Comment by chalmerlowe** [comment]
> @chalmerlowe just pushed the changes based on your comments. Quick question regarding `import test_utils.prefixer` and `import test_utils.retry`, are these internal packages? I was trying to install these to execute the tests on my side but could not figure out what package this is.

I will take a look at the changes!

`test_utils` should be based on this public repo managed by Google:

[](https://github.com/googleapis/python-test-utils/tree/main/test_utils)

It should be possible to install it locally using the following command:

`pip install google-cloud-testutils`

File: tests/system/test_sqlalchemy_bigquery.py [hunk 2]
```diff
@@ -774,3 +774,27 @@ def test_unnest(engine, bigquery_dataset):
...
     assert sorted(r[0] for r in conn.execute(query)) == ["a", "b", "c", "x", "y"]
+
+
+@pytest.mark.skipif(
+    packaging.version.parse(sqlalchemy.__version__) < packaging.version.parse("1.4"),
+    reason="regexp_match support requires version 1.4",
+)
+def test_regexp_match(session, table):
+    results = session.query(table.c.string).where(
+        table.c.string.regexp_match(".*52 St &.*").all()
```
Comment (chalmerlowe) [inline tests/system/test_sqlalchemy_bigquery.py:785]: We are getting error messages/failing tests for this assertion.

The result of a `regexp_match()` function call does not have an `all()` method ... I suspect we need the `.all()` method call to be segregated from the query via a set of closing parenthesis so that it is associated with the overall query. Right now, it is tied to the `regexp_match` and the closing parenthesis are on the following line. I think something closer to this without a closing parens on line 786 might do it.
```suggestion
def test_regexp_match(session, table):
    results = session.query(table.c.string).where(
        table.c.string.regexp_match(".*52 St &.*")).all()
```

**Error message**

```
Traceback (most recent call last):
  File "/tmpfs/src/github/python-bigquery-sqlalchemy/tests/system/test_sqlalchemy_bigquery.py", line 785, in test_regexp_match
    table.c.string.regexp_match(".*52 St &.*").all()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmpfs/src/github/python-bigquery-sqlalchemy/.nox/system-3-11/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 857, in __getattr__
    util.raise_(
  File "/tmpfs/src/github/python-bigquery-sqlalchemy/.nox/system-3-11/lib/python3.11/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
AttributeError: Neither 'BinaryExpression' object nor 'Comparator' object has an attribute 'all'
```

File: tests/system/test_sqlalchemy_bigquery.py [hunk 3]
```diff
@@ -774,3 +774,27 @@ def test_unnest(engine, bigquery_dataset):
...
+
+
+@pytest.mark.skipif(
+    packaging.version.parse(sqlalchemy.__version__) < packaging.version.parse("1.4"),
+    reason="regexp_match support requires version 1.4",
+)
+def test_not_regexp_match(session, table):
+    results = session.query(table.c.string).where(
+        not_(table.c.string.regexp_match("^Barrow St & Hudson St$"))
+    )
```
Comment (chalmerlowe) [inline tests/system/test_sqlalchemy_bigquery.py:798]: Thinking this might do it.

```suggestion
def test_not_regexp_match(session, table):
    results = session.query(table.c.string).where(
        not_(table.c.string.regexp_match("^Barrow St & Hudson St$"))).all()
```

**Review by chalmerlowe — CHANGES_REQUESTED** [review]
Things are coming together... got a bit of quirkiness with the tests we need to figure out.

**Comment by TeddyCr** [comment]
Thanks @chalmerlowe. I was able to run the tests on my end. Tests are passing in my environment ✅

**Comment by chalmerlowe** [comment]
@TeddyCr The coverage tests are failing in kokoro. The coverage doesn't believe the new lines of code are adequately covered by testing.
Let me figure out how I wanna handle that.
I will provide some feedback as soon as I can.

**Comment by TeddyCr** [comment]
> @TeddyCr The coverage tests are failing in kokoro. The coverage doesn't believe the new lines of code are adequately covered by testing.
> Let me figure out how I wanna handle that.
> I will provide some feedback as soon as I can.

@chalmerlowe let me know how you want to proceed

**Comment by chalmerlowe** [comment]
@TeddyCr 
I have been looking at some options and conferring with some colleagues. I am hoping to have an answer by Monday.

**Comment by TeddyCr** [comment]
> @TeddyCr
> I have been looking at some options and conferring with some colleagues. I am hoping to have an answer by Monday.

Thank you 🙏

**Comment by chalmerlowe** [comment]
@TeddyCr Had a headache yesterday. Should have something for you today.

**Comment by chalmerlowe** [comment]
@TeddyCr 
I suggested a few items for your consideration. Those suggestions might be showing up on your original commits on your repo as opposed to being present in this thread. Sorry about that. Computers are hard. : )

**Comment by chalmerlowe** [comment]
@TeddyCr 
I have some time today. I am gonna try to migrate my suggested changes into this repo along with your changes.
If the tests pass, I plan on trying to close out this PR.

**Comment by TeddyCr** [comment]
> 

Sounds good, thanks @chalmerlowe let me know if you need me to do anything.

**Comment by TeddyCr** [comment]
> @TeddyCr
> 
> Nothing I do seems to help. I have been trying to add edits, etc to this PR, but I am failing. All my suggestions go straight to your original repo as suggestions from me.
> 
> Can you please go to your repo and accept those changes to the PR and see if we can get them pulled over to this repo that way? Here is a link:
> 
> [TeddyCr#1](https://github.com/TeddyCr/python-bigquery-sqlalchemy/pull/1)

Just merged the changes to my PR 🙂

**Review by chalmerlowe — APPROVED** [review]
LGTM

**Comment by chalmerlowe** [comment]
@TeddyCr Thank you for your time and effort. And thanks for your patience.
This should be included in the next release.
